precompile/testutils/test_precompile.go

```diff
@@ -69,18 +98,76 @@ func (test PrecompileTest) Run(t *testing.T, module modules.Module, state contra
-		ret, remainingGas, err := module.Contract.Run(accesibleState, test.Caller, contractAddress, input, test.SuppliedGas, test.ReadOnly)
-		if len(test.ExpectedErr) != 0 {
-			require.ErrorContains(t, err, test.ExpectedErr)
-		} else {
-			require.NoError(t, err)
-		}
-		require.Equal(t, uint64(0), remainingGas)
-		require.Equal(t, test.ExpectedRes, ret)
+	return PrecompileRunparams{
+		AccessibleState: accesibleState,
+		Caller:          test.Caller,
+		ContractAddress: contractAddress,
+		Input:           input,
+		SuppliedGas:     test.SuppliedGas,
+		ReadOnly:        test.ReadOnly,
+	}
+}
+
+func (test PrecompileTest) Bench(b *testing.B, module modules.Module, state contract.StateDB) {
+	runParams := test.setup(b, module, state)
+
+	if runParams.Input == nil {
+		b.Skip("Skipping precompile benchmark due to nil input (used for configuration tests)")
 	}
 
+	stateDB := runParams.AccessibleState.GetStateDB()
+	snapshot := stateDB.Snapshot()
+
+	ret, remainingGas, err := module.Contract.Run(runParams.AccessibleState, runParams.Caller, runParams.ContractAddress, runParams.Input, runParams.SuppliedGas, runParams.ReadOnly)
+	if len(test.ExpectedErr) != 0 {
+		require.ErrorContains(b, err, test.ExpectedErr)
+	} else {
+		require.NoError(b, err)
+	}
+	require.Equal(b, uint64(0), remainingGas)
+	require.Equal(b, test.ExpectedRes, ret)
+
 	if test.AfterHook != nil {
-		test.AfterHook(t, state)
+		test.AfterHook(b, state)
+	}
+
+	b.ReportAllocs()
+	start := time.Now()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		// Revert to the previous snapshot and take a new snapshot, so we can reset the state after execution
+		stateDB.RevertToSnapshot(snapshot)
+		snapshot = stateDB.Snapshot()
+
+		// Ignore return values for benchmark
+		_, _, _ = module.Contract.Run(runParams.AccessibleState, runParams.Caller, runParams.ContractAddress, runParams.Input, runParams.SuppliedGas, runParams.ReadOnly)
+	}
+	b.StopTimer()
+
+	elapsed := uint64(time.Since(start))
+	if elapsed < 1 {
+		elapsed = 1
+	}
+	gasUsed := runParams.SuppliedGas * uint64(b.N)
+	b.ReportMetric(float64(runParams.SuppliedGas), "gas/op")
+	// Keep it as uint64, multiply 100 to get two digit float later
+	mgasps := (100 * 1000 * gasUsed) / elapsed
+	b.ReportMetric(float64(mgasps)/100, "mgas/s")
+
+	// Execute the test one final time to ensure that if our RevertToSnapshot logic breaks such that each run is actually failing or resulting in unexpected behavior
+	// the benchmark should catch the error here.
+	stateDB.RevertToSnapshot(snapshot)
+	ret, remainingGas, err = module.Contract.Run(runParams.AccessibleState, runParams.Caller, runParams.ContractAddress, runParams.Input, runParams.SuppliedGas, runParams.ReadOnly)
+	if len(test.ExpectedErr) != 0 {
+		require.ErrorContains(b, err, test.ExpectedErr)
+	} else {
+		require.NoError(b, err)
+	}
+	require.Equal(b, uint64(0), remainingGas)
+	require.Equal(b, test.ExpectedRes, ret)
+
+	if test.AfterHook != nil {
+		test.AfterHook(b, state)
 	}
```

Comment:
hm this sounds like another unit test to me rather than a benchmark

Reply:
As per the comment, this is intended to ensure that the benchmark is passing and not benchmarking an execution that is silently failing, which we would want to catch.